add docker-bench-security results parser, add parser tests

drJabber opened the pull request:
Added OWASP [docker-bench-security](https://github.com/docker/docker-bench-security) results parser, add parser tests
docker-bench-security is a script that make tests based on [CIS Docker Benchmark](https://www.cisecurity.org/benchmark/docker/)

damiencarol commented on unittests/tools/test_dockerbench_parser.py:
```diff
@@ -0,0 +1,30 @@
+        )
+        parser = DockerBenchParser()
+        findings = parser.get_findings(testfile, Test())
+        self.assertTrue(len(findings) == 50)
```
Please add more check on the findings

drJabber replied:
@damiencarol , thank you,  added some extra assertions in this test

drJabber replied:
@damiencarol , fixed failed unit tests.

damiencarol commented on dojo/tools/dockerbench/parser.py:
```diff
@@ -0,0 +1,115 @@
+        description += '\n'
+        description += 'Actual value: {}\n'.format(vuln['actual_value'])
+
+    mitigation = vuln.get('remediation', '')
```
```suggestion
    mitigation = vuln.get('remediation')
```
Mitigation attribute is not mandatory

drJabber replied:
Yes, really, thank you

damiencarol commented on dojo/tools/dockerbench/parser.py:
```diff
@@ -0,0 +1,115 @@
+
+    vuln_id_from_tool = test_number
+
+    finding = Finding(title=title,
+                      test=test,
+                      description=description,
+                      severity=severity,
+                      mitigation=mitigation,
+                      vuln_id_from_tool=vuln_id_from_tool,
+                      static_finding=True,
+                      dynamic_finding=False)
```
The date attribute is missing, you have the date in the `end` attribute of the report.

drJabber replied:
thank you, test date is important. but it does not contain TZ. 
so, I decode it as a local timestamp - datetime.fromtimestamp

damiencarol replied:
Yes, you can do that

damiencarol commented on dojo/tools/dockerbench/parser.py:
```diff
@@ -0,0 +1,115 @@
+    if 'expected_result' in vuln and vuln['expected_result'] != '':
+        description += '\n'
+        description += 'Expected result: {}\n'.format(vuln['expected_result'])
+    if 'actual_value' in vuln and vuln['actual_value'] != '':
```
```suggestion
    if vuln.get('actual_value'):
```

drJabber replied:
Ok, the same as remediation

damiencarol commented on dojo/tools/dockerbench/parser.py:
```diff
@@ -0,0 +1,115 @@
+    elif status.upper() == 'NOTE' and '(Manual)' not in reason:
+        severity = 'Info'
+    else:
+        return None
```
this is not good, severity is mandatory.

drJabber replied:
Thank you, but severity is mandatory when we add the item to result findings. 
Here I dont assign None to severity attribute of finding, here I return None from get_item function (e.g when status is PASS)- I dont create finding in this situation and dont return it to Dojo in get_findings result. 
So I think I can return None when the docker-security-bench result item is not a finding.

damiencarol commented on dojo/tools/dockerbench/parser.py:
```diff
@@ -0,0 +1,115 @@
+    else:
+        return None
+
+    test_number = vuln.get('id', 'Test number not found')
```
I think we should consider that `id` is require for this parser and organize the code like this.

drJabber replied:
I agree, but I think code should not crash when there no id (user could edit file manually),
i tried to fix and now:
vuln_id_from_tool will be id or None, 
and title will be "<id> - <desc>", or "No test number - <desc>"

damiencarol replied:
this is good, will check again the code

damiencarol reviewed (changes requested):
It's a good first shot.
Some stuffs are missing (like severity)
Also you should add de-duplication default configuration in dojo/settings/dist.settings.py

drJabber commented:
> It's a good first shot. Some stuffs are missing (like severity) Also you should add de-duplication default configuration in dojo/settings/dist.settings.py

Thank you, @damiencarol , tried to fix all, also added dedupe config to dist.settings.py

damiencarol commented on dojo/settings/settings.dist.py:
```diff
@@ -1123,6 +1123,7 @@ def saml2_attrib_map_format(dict):
     'Hydra Scan': ['title', 'description'],
     'DrHeader JSON Importer': ['title', 'description'],
     'PWN SAST': ['title', 'description'],
+    'docker-bench-security Scan': ['title', 'description'],
```
I think we should use vuln_id_from_tool here

drJabber replied:
ok, thank you, t think you mean - only vuln_id_from_tool?

damiencarol replied:
yep, but I saw you fixed that

damiencarol commented:
@drJabber  seems good to me but I think you can improve it.
1. more checks in the unit tests (testing more attributes)
2. maybe evaluate switching to use unique_id_from_tool attribute instead of vuln_id_from_tool

drJabber commented:
@damiencarol , thank you, I feel the code is getting better and better :)
>     1. more checks in the unit tests (testing more attributes)

added extra attributes to unit tests - vuln_id_from_tool, severity, title, description, mitigation. I thing here are enough attributes for test, because there are no more data provided by docker-bench-security

>     2. maybe evaluate switching to use unique_id_from_tool attribute instead of vuln_id_from_tool

added unique_id_from_tool to parser and dedupe config, 
but there are no more unique attributes than "id" in docker-bench-security results, because this id is the section number in CIS Docker benchmark document, so unique_id_from_tool==vuln_id_from_tool=="id"

damiencarol commented:
this is good add-on. I think the PR is ready

drJabber commented:
Thank you, @damiencarol 
> Just remove the part for vuln_id_from_tool, it could be confusing
Substituted vuln_id_from_tool by unique id from tool

Also fixed unittests

StefanFl reviewed (changes requested):
Please add the parser in `docs/content/en/integration/parsers.md`.

StefanFl commented:
One tiny change request. Great addition, thank you @drJabber

drJabber commented:
@StefanFl , thank you

> Please add the parser in `docs/content/en/integration/parsers.md`.

Added record in this document

Also fixed unittest

damiencarol commented:
@StefanFl it's good now, could you review it again?

drJabber commented:
Sorry, modified unittest again - incorrectly manually counted issues.

drJabber commented:
Hi guys, I invoked unit tests in my local environment, as written in test guide - and there are no issues on it, could you please review it again?

Maffooch reviewed (approved):
Wow 4 approvals from @damiencarol must be a solid PR!

@StefanFl leaving the merge to you if all looks good

damiencarol commented:
> Wow 4 approvals from @damiencarol must be a solid PR!
> 
> @StefanFl leaving the merge to you if all looks good

@drJabber added many changes to improve quality, this explain the multiple reviews :D

damiencarol commented:
> Wow 4 approvals from @damiencarol must be a solid PR!
> 
> @StefanFl leaving the merge to you if all looks good

@drJabber added many changes to improve quality, this explain the multiple reviews :D

drJabber commented:
Thank you, gentlemen :)
